stubs/WebOb/webob/request.pyi

```diff
@@ -0,0 +1,244 @@
+    no_transform: bool
+    max_age: int
+
+class _NoDefault: ...
+
+NoDefault: _NoDefault
+
+class BaseRequest:
+    request_body_tempfile_limit: ClassVar[int]
+    method: _HTTPMethod
+    def __init__(self, environ: dict[str, Any], **kw: Any) -> None: ...
+    def encget(self, key: str, default: Any = ..., encattr: str | None = ...) -> Any: ...
+    def encset(self, key: str, val: Any, encattr: str | None = ...) -> None: ...
+    @property
+    def charset(self) -> str | None: ...
+    def decode(self, charset: str | None = ..., errors: str = ...) -> Self: ...
+    @property
+    def body_file(self) -> SupportsRead[bytes]: ...
+    @body_file.setter
+    def body_file(self, value: SupportsRead[bytes]) -> None: ...
+    content_length: int | None
+    body_file_raw: SupportsRead[bytes]
+    is_body_seekable: bool
+    @property
+    def body_file_seekable(self) -> IO[bytes]: ...
+    url_encoding: str
+    @property
+    def scheme(self) -> str | None: ...
+    @scheme.setter
+    def scheme(self, value: str | None) -> None: ...
+    @property
+    def http_version(self) -> str | None: ...
+    @http_version.setter
+    def http_version(self, value: str | None) -> None: ...
+    remote_user: str | None
+    remote_host: str | None
+    remote_addr: str | None
+    query_string: str
+    @property
+    def server_name(self) -> str | None: ...
+    @server_name.setter
+    def server_name(self, value: str | None) -> None: ...
+    @property
+    def server_port(self) -> int | None: ...
+    @server_port.setter
+    def server_port(self, value: int | None) -> None: ...
+    script_name: str
+    @property
+    def path_info(self) -> str | None: ...
+    @path_info.setter
+    def path_info(self, value: str | None) -> None: ...
+    uscript_name: str  # bw compat
+    upath_info = path_info  # bw compat
+    content_type: str | None
+    @property
+    def headers(self) -> EnvironHeaders: ...
+    @headers.setter
+    def headers(self, value: SupportsItems[str, str] | Iterable[tuple[str, str]]) -> None: ...
+    @property
+    def client_addr(self) -> str | None: ...
+    @property
+    def host_port(self) -> str: ...
+    @property
+    def host_url(self) -> str: ...
+    @property
+    def application_url(self) -> str: ...
+    @property
+    def path_url(self) -> str: ...
+    @property
+    def path(self) -> str: ...
+    @property
+    def path_qs(self) -> str: ...
+    @property
+    def url(self) -> str: ...
+    def relative_url(self, other_url: str, to_application: bool = ...) -> str: ...
+    def path_info_pop(self, pattern: Pattern[str] | None = ...) -> str | None: ...
+    def path_info_peek(self) -> str | None: ...
+    urlvars: dict[str, str]
+    urlargs: tuple[str]
+    @property
+    def is_xhr(self) -> bool: ...
+    host: str
+    @property
+    def domain(self) -> str: ...
+    body: bytes
+    json: Any
+    json_body: Any
+    text: str
+    @property
+    def POST(self) -> MultiDict[str, str] | NoVars: ...
```

Comment:
Technically this should be `MultiDict[str, str | cgi.FieldStorage] | NoVars`. In practice having to deal with `isinstance` checks on every value you get out of `POST` is fairly annoying, but it is also the only way to be fully type safe, so I'm not sure what the best answer is here. Having `NoVars` in the type union is annoying as well, since usually you don't really care whether it's `NoVars` or `MultiDict` in a request, since `NoVars` will behave just like an empty `MultiDict`. Sometimes it's useful to distinguish between an empty POST and this request can't even have POST parameters to begin with, but `Request.method` should be sufficient for that...

TLDR: Should I prioritize type safety or ease of use? This is also kind of a use-case where `AnyOf` would be nice to have. It's a nice band-aid solution for APIs that weren't really thought out well enough in terms of type safety, without having to throw type safety completely out of the window by returning `Any`.

Reply:
I ended up switching to the type safe version that does take `cgi.FieldStorage` into account, but it uses a dummy subclass to let static analyzers know that `FieldStorage.file` and `FieldStorage.filename` are always set, to make it slightly less annoying to use. webob interally subclasses FieldStorage to slightly adjust the behavior, but since it does not change the way people interface with the object, it should not matter, as long as what we return passes `isinstance(fs, cgi.FieldStorage)`.